### PR TITLE
Closes #77 | Create dataset loader for CC3M-35L

### LINKED ISSUE #77: Create dataset loader for CC3M-35L
Dataloader name: `cc3m_35l/cc3m_35l.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?cc3m_35l

|     Dataset| cc3m_35l |
|-------------|---|
| Description | CC3M-35L is created by translating Conceptual Captions 3M (Sharma et al., 2018), originally in English, to the other 34 languages using Google’s machine translation API. |
| Subsets | fil, ind, tha, vie |
| Languages | fil, ind, tha, vie |
| Tasks | Image-to-Text Generation |
| License | Creative Commons Attribution 4.0 (cc-by-4.0) |
| Homepage | CC3M-35L: https://google.github.io/crossmodal-3600/; CC3M: https://ai.google.com/research/ConceptualCaptions/download|
| HF URL | - |
| Paper URL | https://aclanthology.org/2022.emnlp-main.45/ |

### PR DESCRIPTION
**Pull request by IvanHalimP**
Please name your PR after the issue it closes. You can use the following line: "Closes #ISSUE-NUMBER" where you replace the ISSUE-NUMBER with the one corresponding to your dataset.
Closes #77 
### Checkbox
- [ ] Confirm that this PR is linked to the dataset issue.
- [ ] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [ ] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [ ] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [ ] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [ ] Confirm dataloader script works with `datasets.load_dataset` function.
- [ ] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

**Comment by IvanHalimP** [comment]
There are some things I need to mention here.
1. The original captions need to be downloaded using one's Google account and loaded locally.
2. I tried processing this without parallelization by default. It takes days to finish (not even finished by the time I committed this)
3. I tried running the validation set only. Still takes too long.
4. Implemented ProcessPoolExecutor and ThreadPoolExecutor to make things quicker. Works as intended but require a lot of memory. I only have 64GB of RAM available and it is not enough to load the train set.
5. The URLs given in the dataset are not guaranteed to be downloadable. Sometimes the connection/security certificate and other factors may raise exceptions. Hence the total samples may vary from time to time.

**Comment by SamuelCahyawijaya** [comment]
== Note ==
Home page: https://ai.google.com/research/ConceptualCaptions/download
Training TSV: https://storage.cloud.google.com/gcc-data/Train/GCC-training.tsv
Validation TSV: https://storage.cloud.google.com/gcc-data/Validation/GCC-1.1.0-Validation.tsv

**Comment by jen-santoso** [comment]
> I haven't tested it with train set, but the validation set is already ok. Perhaps, not downloading the images is a good idea to speed things up?

@IvanHalimP Could you please provide the output of the unit tests in the PR? Just copy and paste here, even if it's just for validation set. It is required, at least in the checklist. Please finish this first, then I will continue with the review

**Comment by IvanHalimP** [comment]
```
(env-seacrowd) halimp-i@xxxdata01:/proj_disk/halimp-i/code/seacrowd-datahub$ python -m tests.test_seacrowd seacrowd/sea_datasets/cc3m_35l/cc3m_35l.py --subset_id=cc3m_35l_id --data_dir=/proj_disk/halimp-i/dataset
INFO:__main__:args: Namespace(path='seacrowd/sea_datasets/cc3m_35l/cc3m_35l.py', schema=None, subset_id='cc3m_35l_id', data_dir='/proj_disk/halimp-i/dataset', use_auth_token=None)
INFO:__main__:self.PATH: seacrowd/sea_datasets/cc3m_35l/cc3m_35l.py
INFO:__main__:self.SUBSET_ID: cc3m_35l_id
INFO:__main__:self.SCHEMA: None
INFO:__main__:self.DATA_DIR: /proj_disk/halimp-i/dataset
INFO:__main__:Checking for _SUPPORTED_TASKS ...
module seacrowd.sea_datasets.cc3m_35l.cc3m_35l
INFO:__main__:Found _SUPPORTED_TASKS=[<Tasks.IMAGE_CAPTIONING: 'IC'>]
INFO:__main__:_SUPPORTED_TASKS implies _MAPPED_SCHEMAS={'IMTEXT'}
INFO:__main__:schemas_to_check: {'IMTEXT'}
INFO:__main__:Checking load_dataset with config name cc3m_35l_id_source
/user_base/halimp-i/anaconda3/envs/env-seacrowd/lib/python3.10/site-packages/datasets/load.py:2088: FutureWarning: 'use_auth_token' was deprecated in favor of 'token' in version 2.14.0 and will be removed in 3.0.0.
You can remove this warning by passing 'token=<use_auth_token>' instead.
  warnings.warn(
Downloading data: 1.14MB [00:00, 3.59MB/s]

...[downloading images... many many lines...]...

Generating train split: 3318333 examples [03:25, 16164.15 examples/s]
Generating validation split: 15839 examples [00:01, 15573.88 examples/s]
INFO:__main__:Checking load_dataset with config name cc3m_35l_id_seacrowd_imtext
/user_base/halimp-i/anaconda3/envs/env-seacrowd/lib/python3.10/site-packages/datasets/load.py:2088: FutureWarning: 'use_auth_token' was deprecated in favor of 'token' in version 2.14.0 and will be removed in 3.0.0.
You can remove this warning by passing 'token=<use_auth_token>' instead.
  warnings.warn(
Downloading data: 1.14MB [00:00, 4.27MB/s]
Downloading data: 1.14MB [00:00, 3.91MB/s]
Downloading data: 1.14MB [00:01, 702kB/s]
Downloading data: 1.14MB [00:00, 1.76MB/s]
Generating train split: 3318333 examples [03:13, 17166.20 examples/s]
Generating validation split: 15839 examples [00:00, 16465.90 examples/s]
INFO:__main__:Dataset sample [source]
{'id': '0', 'image_paths': None, 'src_lang': 'en', 'caption_tokenized': 'a very typical bus station', 'trg_lang': 'id', 'translation_tokenized': None, 'backtranslation_tokenized': None}
INFO:__main__:Dataset sample [seacrowd_imtext]
{'id': '0', 'image_paths': [None], 'texts': None, 'metadata': {'context': None, 'labels': None}}
INFO:__main__:Checking global ID uniqueness
INFO:__main__:Found 15839 unique IDs
INFO:__main__:Gathering schema statistics
INFO:__main__:Gathering schema statistics
train
==========
id: 3318333
image_paths: 3318333
metadata: 6636666
texts: 12

validation
==========
id: 15839
image_paths: 15839
metadata: 31678
texts: 19

.
----------------------------------------------------------------------
Ran 1 test in 15892.511s

OK
```

I need to change the code back into no parallelization since it kept being killed by the OS.

**Comment by holylovenia** [comment]
Hi @IvanHalimP, the reviewers (@SamuelCahyawijaya & @jen-santoso) informed me that this dataloader is quite complex and that they would like to discuss it with you sometime to sort out the confusion. However, their hands are tied with work/study duties right now, so they will inform you and arrange it with you at a later time. I hope this message clears up as to why this dataloader hasn't been reviewed yet.

**Comment by IvanHalimP** [comment]
> Hi @IvanHalimP, the reviewers (@SamuelCahyawijaya & @jen-santoso) informed me that this dataloader is quite complex and that they would like to discuss it with you sometime to sort out the confusion. However, their hands are tied with work/study duties right now, so they will inform you and arrange it with you at a later time. I hope this message clears up as to why this dataloader hasn't been reviewed yet.

Hi @holylovenia , thanks for the update.  That's totally understandable, I'm pretty much at the same situation here.

File: seacrowd/sea_datasets/cc3m_35l/cc3m_35l.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+            
+                
+            #for line in train_caption_targets[:100]: # this was for debugging only
+            for line in train_caption_targets:
+                res = self.fill_img_path(gcc_val_df, line)
+                train_exceptions.extend(res[1])
+                gcc_train_df.update(res[0])
+
+        # clean the memory
+        train_caption_targets = []
```
Comment (jen-santoso) [inline seacrowd/sea_datasets/cc3m_35l/cc3m_35l.py:196]: Regarding this part, initially I had trouble with the extremely long loading time (more than days) as you reported. @SamuelCahyawijaya had tried to modify the code to support mapping (CMIIW) so the images are downloaded only when the data is fetched. The mapping (string matching part) was improved so it was not OOM. As a result the loading became much faster (downloading is a few hours at most, and loading takes minutes). Note that this dataloader should be run with `streaming=True` on the local test (not from the test module).

So, @IvanHalimP, would you mind if we update this part? In the meantime, could you clear the minor nits for a while?

**Review by jen-santoso — CHANGES_REQUESTED** [review]
@IvanHalimP Thank you for the implementation and the reports. 
Thanks for waiting, we had a bit of difficulty in accelerating the data loading process.

The dataloader for now loads well, with a few modifications from our side. (kudos to @SamuelCahyawijaya for the `streaming` setting and string mapping of image/filename)
Aside of the minor nits I'd pointed out (please fix this), would you mind if we provide the update on the acceleration of the data loader? This requires us to update the `_split_generators`.

File: seacrowd/sea_datasets/cc3m_35l/cc3m_35l.py (hunk 1 above)
Comment (IvanHalimP) [reply seacrowd/sea_datasets/cc3m_35l/cc3m_35l.py:196]: Sure thing!
Thank you @SamuelCahyawijaya for your help.

**Comment by holylovenia** [comment]
Friendly reminder for @SamuelCahyawijaya and @jen-santoso to follow up! I'm just making sure this PR did not get buried. 🙏

**Review by SamuelCahyawijaya — APPROVED** [review]
@IvanHalimP  @jen-santoso: LGTM!!! Finally!! 🚀🚀🚀

**Comment by jen-santoso** [comment]
Yes! Checked the PR, LGTM!
@SamuelCahyawijaya so, just curious, but in the end how do we run the code in seacrowd test with `streaming=True` in the end?

**Review by jen-santoso — APPROVED** [review]

**Comment by SamuelCahyawijaya** [comment]
@jen-santoso : I didn't, I just test it manually using `datasets.load_dataset()`.
